File: code_samples/section7/problems/problem7_2/problem7_2.js
/**
 * groupAnagrams
 *
 * Purpose:
 *   Groups strings that are anagrams of each other.
 *
 * Definition:
 *   Two strings are anagrams if they contain the same characters with the same
 *   frequencies, possibly in a different order.
 *
 * Parameters:
 *   strs - Array of input strings.
 *
 * Returns:
 *   An array of groups, where each group is an array of strings that are
 *   anagrams of each other.
 *
 * Approach:
 *   - Use a Map where:
 *       key   = canonical representation of the string
 *       value = array of original strings that match that key
 *
 *   - Canonical key strategy:
 *       Split the string into characters, sort them, and re-join.
 *       All anagrams produce the same sorted string.
 *       Example:
 *         "eat" -> ["e","a","t"] -> ["a","e","t"] -> "aet"
 *         "tea" -> "aet"  (same key → same group)
 *
 * Complexity:
 *   Let:
 *     n = number of strings
 *     k = average string length
 *
 *   - Time:  O(n · k log k)
 *       Sorting each string dominates the cost.
 *   - Space: O(n · k)
 *       For storing keys and grouped strings.
 *
 * Notes:
 *   - Map preserves insertion order of keys, but the order of groups
 *     still depends on the first time each anagram key is encountered.
 *   - Group order and item order are allowed to vary for this problem.
 */
function groupAnagrams(strs) {
    // Map: sorted-string key -> array of original strings
    const map = new Map();

    // Process each string in the input array
    for (const s of strs) {
        // Create the canonical anagram key by sorting characters
        const key = s.split("").sort().join("");

        // Initialize a new group if this key hasn't been seen before
        if (!map.has(key)) {
            map.set(key, []);
        }

        // Add the original string to its anagram group
        map.get(key).push(s);
    }

    // Convert the map's values (arrays of strings) into a plain array
    // to match the expected return type.
    return Array.from(map.values());
}

// ============================
// Tests (top-level)
// ============================
//
// Simple console-based test harness demonstrating usage.
//

const input = ["eat", "tea", "tan", "ate", "nat", "bat"];

console.log("=== Test: groupAnagrams ===\n");
console.log("Input:", input, "\n");

// Run the function under test
const groups = groupAnagrams(input);

// Print each anagram group
console.log("Anagram Groups:");
for (const g of groups) {
    console.log("  [" + g.join(", ") + "]");
}

// Expected output guidance (ordering may differ)
console.log("\nExpected groups (order may vary):");
console.log("  [eat, tea, ate]");
console.log("  [tan, nat]");
console.log("  [bat]");

File: code_samples/section7/problems/problem7_2/problem7_2.java
package code_samples.section7.problems.problem7_2;

import java.util.*;

/**
 * problem7_2
 *
 * Solves the classic "Group Anagrams" problem.
 *
 * Given an array of strings, group the strings that are anagrams of each other.
 * Two strings are anagrams if they contain the same characters with the same
 * frequencies, possibly in a different order.
 *
 * This implementation uses:
 *  - HashMap for grouping by a canonical key
 *  - Sorted-character strings as the anagram key
 */
public class problem7_2 {

    /**
     * groupAnagrams
     *
     * @param strs Array of input strings to be grouped.
     * @return A list of anagram groups, where each group is a list of strings.
     *
     * Approach:
     *   - For each string:
     *       1) Convert it to a char array.
     *       2) Sort the characters.
     *       3) Convert the sorted characters back into a string.
     *          This sorted string becomes the canonical key for all its anagrams.
     *       4) Insert the original string into a map keyed by this canonical form.
     *
     * Example:
     *   "eat" -> ['e','a','t'] -> ['a','e','t'] -> "aet"
     *   "tea" -> "aet"  (same key → same group)
     *
     * Complexity:
     *   Let:
     *     n = number of strings
     *     k = average length of a string
     *
     *   - Time:  O(n · k log k)
     *       Sorting each string dominates the cost.
     *   - Space: O(n · k)
     *       For storing keys and grouped strings.
     *
     * Notes:
     *   - The order of groups and the order of strings within each group
     *     are not guaranteed, because HashMap does not preserve insertion order.
     */
    List<List<String>> groupAnagrams(String[] strs) {
        // Map: sorted string (anagram key) -> list of original strings
        Map<String, List<String>> map = new HashMap<>();

        // Process each input string
        for (String s : strs) {
            // Convert the string into a mutable char array
            char[] chars = s.toCharArray();

            // Sort characters alphabetically to form the canonical key
            Arrays.sort(chars);

            // Convert sorted characters back into a string key
            String key = new String(chars);

            // Fetch the existing group for this key, or create it if absent,
            // then add the original string to that group.
            map.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        // Return all grouped anagram lists as a List<List<String>>
        return new ArrayList<>(map.values());
    }

    // ================================
    // Test Code
    // ================================

    /**
     * main
     *
     * Simple console-based test harness to demonstrate groupAnagrams().
     * This is not a formal unit test (e.g., JUnit), but it clearly shows
     * expected behavior and output.
     */
    public static void main(String[] args) {
        problem7_2 sol = new problem7_2();

        // Input words to group
        String[] input = {"eat", "tea", "tan", "ate", "nat", "bat"};

        System.out.println("=== Test: groupAnagrams ===\n");
        System.out.println("Input: " + Arrays.toString(input) + "\n");

        // Run the solution
        List<List<String>> groups = sol.groupAnagrams(input);

        // Print the resulting groups
        System.out.println("Anagram Groups:");
        for (List<String> group : groups) {
            System.out.println("  " + group);
        }

        // Expected output guidance (ordering may differ)
        System.out.println("\nExpected groups (order may vary):");
        System.out.println("  [eat, tea, ate]");
        System.out.println("  [tan, nat]");
        System.out.println("  [bat]");
    }
}
